In [1]:
import sys
import os
sys.path.append(os.path.abspath('../scripts'))

# Get the current directory
current_dir = os.getcwd()
print(current_dir)

# Get the parent directory
parent_dir = os.path.dirname(current_dir)
print(parent_dir)

# Insert the path to the parent directory
sys.path.insert(0, parent_dir)

from scripts.load_data import load_data_using_sqlalchemy

e:\10Academy\Tellco_DA\notebooks
e:\10Academy\Tellco_DA
DB_HOST: localhost
DB_PORT: 5432
DB_NAME: Telecom_db
DB_USER: postgres
DB_PASSWORD: root


database connection and loading data

In [2]:
# Define your SQL query
query = "SELECT * FROM xdr_data;"  # Replace with your actual table name

# Load data from PostgreSQL using SQLAlchemy
df = load_data_using_sqlalchemy(query)

# Display the first few rows of the dataframe
if df is not None:
    print("Successfully loaded the data")
else:
    print("Failed to load data.")

Successfully loaded the data


In [6]:
# Import the OverviewAnalyzer class from the UserOverview_Analysis module
from scripts.UserOverview_Analysis import OverviewAnalyzer

# Initialize the OverviewAnalyzer with a dataset and the column of interest
# - 'df': The DataFrame containing the telecom dataset
# - 'HTTP DL (Bytes)': The column representing HTTP download data in bytes

analyzer = OverviewAnalyzer(df, 'HTTP DL (Bytes)')

In [7]:
# Use the OverviewAnalyzer to compute a table of missing values for the dataset
missing_values_df = analyzer.missing_values_table(df)

# Print the header for missing values report
print("Missing Values in df:")

# Display the missing values DataFrame
print(missing_values_df)

Your selected dataframe has 55 columns.
There are 41 columns that have missing values.
Missing Values in df:
                                          Missing Values  % of Total Values  \
Nb of sec with 37500B < Vol UL                    130254               86.8   
Nb of sec with 6250B < Vol UL < 37500B            111843               74.6   
Nb of sec with 125000B < Vol DL                    97538               65.0   
TCP UL Retrans. Vol (Bytes)                        96649               64.4   
Nb of sec with 31250B < Vol DL < 125000B           93586               62.4   
Nb of sec with 1250B < Vol UL < 6250B              92894               61.9   
Nb of sec with 6250B < Vol DL < 31250B             88317               58.9   
TCP DL Retrans. Vol (Bytes)                        88146               58.8   
HTTP UL (Bytes)                                    81810               54.5   
HTTP DL (Bytes)                                    81474               54.3   
Avg RTT DL (ms)       

In [8]:
# Convert 'HTTP DL (Bytes)' from bytes to megabytes and store in new column 'x'
df['x'] = analyzer.convert_bytes_to_megabytes(df, 'HTTP DL (Bytes)')  
# Print the updated DataFrame with the new 'x' column
print(df)

           Bearer Id            Start  Start ms              End  End ms  \
0       1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1       1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2       1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3       1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4       1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   
...              ...              ...       ...              ...     ...   
149996  7.277826e+18   4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18   4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19   4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19   4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN             None       NaN             None     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0       1823652.0  2.08

In [9]:
# Aggregates user behavior data using the 'aggregate_user_behavior' method from 'analyzer' on the DataFrame 'df' and stores the result in 'user_behavior'
user_behavior = analyzer.aggregate_user_behavior(df)  
# Displays the first few rows of the aggregated user behavior data
print(user_behavior.head())  
# Prints the original DataFrame 'df' to show its content
print(df)

   MSISDN/Number  num_xdr_sessions  total_session_duration  \
0   3.360100e+10                 1                116720.0   
1   3.360100e+10                 1                181230.0   
2   3.360100e+10                 1                134969.0   
3   3.360101e+10                 1                 49878.0   
4   3.360101e+10                 2                 37104.0   

   total_download_data  total_upload_data  social_media_dl  social_media_ul  \
0         8.426375e+08         36053108.0        2206504.0          25631.0   
1         1.207552e+08         36104459.0        2598548.0          62017.0   
2         5.566597e+08         39306820.0        3148004.0          47619.0   
3         4.019932e+08         20327526.0         251469.0          28825.0   
4         1.363130e+09         94280527.0        2861230.0          51312.0   

   youtube_dl  youtube_ul  netflix_dl  ...   google_dl  google_ul   email_dl  \
0  14900201.0   6724347.0  10265105.0  ...   3337123.0  1051882.0   8374

## Summary and Recommendations
The dataset, from a PostgreSQL database, includes 55 columns, with 41 showing missing values, particularly in network traffic and retransmission volumes. Notable columns with missing data include "Nb of sec with 37500B < Vol UL" (86.8% missing), while others like "MSISDN/Number" have minimal missing data. The data provides insights into user behavior, session durations, and data usage across platforms like social media, streaming, and gaming.

The key challenge is missing values in columns like MSISDN/Number and Bearer Id, which should be addressed using imputation. The data is measured in bytes, and converting it to megabytes or gigabytes would enhance clarity.

Recommendations:
Impute Missing Data: Address missing values in critical columns to ensure data completeness.
Convert Units: Change traffic data units from bytes to megabytes or gigabytes.
Review Data Extraction: Improve data capture and extraction processes to minimize missing data.
Enhance Analysis: Use data transformations to improve the interpretability of user behavior patterns.